src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs

```diff
@@ -434,6 +434,38 @@ public void VerifyCompilerAssemblySimpleNames()
                 allReferenceSimpleNames.Free();
             }
         }
+
+        [Fact]
+        public void AssemblyLoadingInNonDefaultContext_AnalyzerReferencesSystemCollectionsImmutable()
+        {
+            // Create a separate ALC as the compiler context, load the compiler assembly and a modified version of S.C.I into it,
+            // then use that to load and run `AssemblyLoadingInNonDefaultContextHelper` below. We expect the analyzer running in
+            // its own `DirectoryLoadContext` would use the bogus S.C.I loaded in the compiler load context instead of the real one
+            // in the default context.
+            var compilerContext = new System.Runtime.Loader.AssemblyLoadContext("compilerContext");
+            _ = compilerContext.LoadFromAssemblyPath(_testFixture.UserSystemCollectionsImmutable.Path);
+            _ = compilerContext.LoadFromAssemblyPath(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly.Location);
+
+            var testAssembly = compilerContext.LoadFromAssemblyPath(typeof(DefaultAnalyzerAssemblyLoaderTests).GetTypeInfo().Assembly.Location);
+            var testObject = testAssembly.CreateInstance(typeof(DefaultAnalyzerAssemblyLoaderTests).FullName!,
+                ignoreCase: false, BindingFlags.Default, binder: null, args: new object[] { _output, _testFixture }, null, null)!;
+
+            StringBuilder sb = new();
+            testObject.GetType().GetMethod(nameof(AssemblyLoadingInNonDefaultContextHelper), BindingFlags.Instance | BindingFlags.NonPublic)!.Invoke(testObject, new object[] { sb });
+            Assert.Equal("42", sb.ToString());
```

Comment:
Silly question, but did you verify that the test fails when the implementation change in this PR is not present?

Reply:
Yes, the test would fail w/o change in this PR.